sys/include/net/nanocoap_sock.h

```diff
@@ -209,6 +209,55 @@ typedef struct {
     uint8_t blksize;                /**< CoAP blocksize exponent            */
 } coap_block_request_t;
 
+/**
+ * @brief   Request context for separate response
+ */
+typedef struct {
+    sock_udp_ep_t remote;           /**< remote to send response to         */
+#if defined(MODULE_SOCK_AUX_LOCAL) || DOXYGEN
+    sock_udp_ep_t local;            /**< local from which to send response  */
+#endif
+    uint8_t token[COAP_TOKEN_LENGTH_MAX];   /**< request token              */
+    uint8_t tkl;                    /**< request token length               */
+    uint8_t no_response;            /**< no-response bitmap                 */
+} nanocoap_sock_response_ctx_t;
+
+/**
+ * @brief   Prepare the context for a separate response
+ *
+ * This function serializes the CoAP request information so that
+ * a separate response can be generated outside the CoAP handler.
+ *
+ * The CoAP handler should then respond with an empty ACK by calling
+ * @ref coap_build_empty_ack
+ *
+ * @param[out]  ctx     Context information for separate response
+ * @param[in]   pkt     CoAP packet to which the response will be generated
+ * @param[in]   req     Context of the CoAP request
+ */
+void nanocoap_sock_prepare_separate(nanocoap_sock_response_ctx_t *ctx,
+                                    coap_pkt_t *pkt, const coap_request_ctx_t *req);
+
+/**
+ * @brief   Send a separate response to a CoAP request
+ *
+ *  This sends a response to a CoAP request outside the CoAP handler
+ *
+ * @pre     @ref nanocoap_sock_prepare_separate has been called on @p ctx
+ *          inside the CoAP handler
+ *
+ * @param[in]   ctx     Context information about the CoAP request
+ * @param[in]   code    CoAP response code
+ * @param[in]   type    Response type, may be `COAP_TYPE_NON` or `COAP_TYPE_RST`
```

Comment:
The response can also be a CON

Reply:
But I think it cannot be a RST because a RST is empty this it contains no token and is meaningless as a response, given that the empty ACK was already sent

Reply:
> assert(type != COAP_TYPE_CON); /* TODO: add support */

What's the issue?

Reply:
I can't do the retransmission handling in the server thread as CON is blocking in nanoCoAP - I'd need a separate thread / event for that.
To keep things simple for now, I thought this might as well ba added later if the need for it comes up.

Reply:
Ah I just realized this is already using events